fixes issue #552 : Scrollable options tabs.

kianzarrin opened the pull request:
_Fixes_ #552 

- put a `UIScrollablePanel` inside each tab page to make it scroll-able
- Encapsulated `UITabstrip` in `ExtUITabstrip`.
- in each MakeSettings* method, used the new `ExtUITabstrip` class and removed duplicated codes.
 - The scroll-able panel does not work with keybinds panel for some reason, so the keybind panels are not scrollable. there is not too many options there anyway. I suggest we deal with it when we need it.
- The panels are not long enough to require scrollbars yet. Not until I merge in #539 in another PR. So you should not notice any visual difference in the appearance yet. 
- Added static Test class to test the scroll-able panels. just uncomment the following to activate the tests:
```public static void MakeSettings(UIHelperBase helper){//ExtUITabstrip.Test.OnSettingsUI(helper);return;```

kvakvs commented:
The code so far is looking good.

Please do `git checkout -b 552-scrollable-options`
Then `git push -u` and follow the suggestion (whatever your remote in Git is named, origin probably, in that case `git push -u origin 552-scrollable-options`
And then you can have this work in a branch other than `master`.

aubergine10 commented:
@kianzarrin Will #562 affect this PR? If we plan on switching to new config system (eg. similar to Real Time), it would probably negate some of the work done in this PR?

kianzarrin commented:
No @aubergine10. #562 does not influence this PR. This is just encapsulation of UI and it runs completely independently of TMPE. Infact you can copy ExtUITabstrip.cs into an empty mod and I will work just fine.

If anything better encapsulation and removing of duplicate code paves the path to any future modification.

In any case this is blocking my other tasks so I need this in ASAP.

aubergine10 commented on TLM/TLM/Manager/Impl/ExtUITabStrip.cs:
```diff
@@ -0,0 +1,151 @@
+            return tabStrip;
+        }
+
+        public static class Test {
```
This static class should be encapsulated in `#if DEBUG ... #endif` block? I don't see any need for it to be in `RELEASE` or `LABS` builds.

krzychu124 commented on TLM/TLM/Manager/Impl/ExtUITabStrip.cs:
```diff
@@ -0,0 +1,154 @@
+            return verticalScrollbar;
+        }
+
+        private UIScrollablePanel CreateScrollebalePanel(UIPanel panel) {
```
Typo 😉

krzychu124 commented on TLM/TLM/Manager/Impl/ExtUITabStrip.cs:
```diff
@@ -0,0 +1,154 @@
+            return panelHelper;
+        }
+
+        public static ExtUITabstrip Create(UIHelperBase helper_base) {
```
Follow naming style, don't use underscore in variable names.

krzychu124 commented on TLM/TLM/State/Options.cs:
```diff
@@ -115,46 +116,15 @@ public class Options : MonoBehaviour {
-            OptionsKeybindsTab.MakeSettings_Keybinds(tabStrip, tabIndex);
-
-            tabStrip.selectedIndex = 0;
+            //ExtUITabstrip.Test.OnSettingsUI(helper);return;
```
remove or add new preprocessor directive before merge

kianzarrin replied:
I will just remove the commented out code. Its only for testing anyway.

krzychu124 commented on TLM/TLM/Manager/Impl/ExtUITabStrip.cs:
```diff
@@ -0,0 +1,154 @@
+    using ColossalFramework.UI;
+
+    public sealed class ExtUITabstrip : UITabstrip {
+        private UIScrollbar CreateVerticalScrollbar(UIPanel panel, UIScrollablePanel scrollablePanel) {
```
If it's possible move scrollbar more to the right edge of options window

kianzarrin replied:
Its as right as it can be.

krzychu124 commented on TLM/TLM/UI/MainMenu/TimedTrafficLightsButton.cs:
```diff
@@ -1,12 +1,12 @@
         protected override ButtonFunction Function => ButtonFunction.TimedTrafficLights;
 
-        public override string Tooltip => Translation.Menu.Get("Tooltip:Timed traffic lights");
+        public override string Tooltip => Translation.Menu.Get("Tooltip:Timed traffic lights") + "\n" + Translation.Menu.Get("Tooltip.Keybinds:Timed traffic lights");
```
We don't have that language key yet 😕

kianzarrin replied:
Don't know how it got here. I got confused. This belongs to another review.

krzychu124 commented:
I've noticed that checkboxes on Overlays tab are more dense positioned than before

__Before__
![Zrzut ekranu (283)](https://user-images.githubusercontent.com/19638970/69484526-64c5e300-0e34-11ea-8467-190231da031b.png)

__This PR__
![Zrzut ekranu (282)](https://user-images.githubusercontent.com/19638970/69484527-67283d00-0e34-11ea-9249-486ebb2a605a.png)

aubergine10 commented:
Not able to test right now, but just before I forget: Ensure that check boxes can be indented. There's one of the tabs where we indent one or more checkbox options.

kianzarrin commented:
I am ready for the next round of review.

kianzarrin commented:
![Screenshot (100)](https://user-images.githubusercontent.com/26344691/69497901-87b2ce80-0eea-11ea-813d-7b5482e22981.png)
The indent works.
(Just in case you are wondering where are the scroll bars, they have autohide if the panel fits the screen)

aubergine10 commented:
One possible thing to check: Should the extended class it be in the `Impl` folder or the `UI` folder?

kianzarrin commented:
@aubergine10 
> One possible thing to check: Should the extended class it be in the `Impl` folder or the `UI` folder?

Thanks for raising this point. I noticed that all classes that have Ext in them are in manager.impl . but now that I think of it they are also linked by manager factory. 

I have never worked on object oriented programming in a professional context so I have nooo ideaaaa  how to organize these things.

If you want me to move it to somewhere else I'll be more than happy to 😄

aubergine10 commented:
I also am not sure which is best folder. Hopefully @krzychu124 or @kvakvs can advise.

kvakvs commented:
I don't have a strong opinion on this. Don't stress yourself too much over file location.
Use something that looks right today, moving it later is a matter of two clicks.

kianzarrin commented:
As the commit messages suggests, in order to make the changes easier to review  I moved ExtUITabStrip to UI.Helpers folder and namespace in one commit and changed nothing else. Then I fixed the padding issue in separate commit. 

@krzychu124 
> I've noticed that checkboxes on Overlays tab are more dense positioned than before

fixed it. 
![Screenshot (108)](https://user-images.githubusercontent.com/26344691/69596372-0a32af80-100b-11ea-9ddb-0ee9fa215056.png)
